File: crates/dev/http/src/status_overlay.js
// The in-page build-state overlay for `idealyst dev`.
//
// Spliced into the reload script (`RELOAD_SCRIPT` in lib.rs), which feeds
// it every `dev-state` event from the reload stream: the same versioned
// objects `idealyst dev --events-file` writes, projected to what
// describes the build (`dev_events::page::wants`).
//
// Plain JS and CSS on purpose: the overlay has to work when the page's
// wasm is the thing that failed to build — a Rust error must still reach
// the screen. A richer overlay can be an idealyst component once the
// bundle is live; this one owes nothing to it.
//
// It shows two things:
// - a small badge in the corner: what the dev loop is doing (building,
//   with cargo's progress and the current stage) or last did (the tier
//   that applied the save, and how long it took);
// - a panel over the page when a build fails: each rustc error with its
//   file:line and rendered message. Esc or a click outside the message
//   dismisses it; the next successful build or patch clears it.
//
// DOM use is deliberately narrow (createElement, appendChild, textContent,
// style.cssText, setAttribute, addEventListener, attachShadow) so the
// behaviour is testable in node against a small fake DOM
// (`crates/dev/http/tests/status_overlay.rs`). The shadow root keeps the
// app's CSS out of the overlay and the overlay's out of the app.
function idealystStatusOverlay(doc) {
  var FONT = "font:12px/1.4 ui-monospace,SFMono-Regular,Menlo,monospace;";
  var host = null, badge = null, panel = null, list = null, title = null;
  var st = {
    phase: "idle",      // idle | change | patching | building | done | error
    text: "",
    stage: null,
    progress: null,     // {compiled, total, current}
    pending: [],        // error diagnostics of the build in flight
    failure: null,      // {target, error, diagnostics} of the last failed build
    dismissed: false
  };

  function el(tag, css) {
    var e = doc.createElement(tag);
    if (css) e.style.cssText = css;
    return e;
  }

  function mount() {
    if (host) return;
    host = el("idealyst-dev-overlay", "all:initial;");
    host.setAttribute("data-idealyst-dev-overlay", "");
    var root = host.attachShadow ? host.attachShadow({ mode: "open" }) : host;
    badge = el("div",
      "position:fixed;right:8px;bottom:8px;z-index:2147483647;max-width:60vw;" +
      "padding:3px 8px;border-radius:4px;background:#111c;color:#eee;" +
      "white-space:nowrap;overflow:hidden;text-overflow:ellipsis;pointer-events:none;" + FONT);
    badge.setAttribute("data-part", "badge");
    panel = el("div",
      "position:fixed;inset:0;z-index:2147483647;background:#000b;display:none;" +
      "overflow:auto;padding:5vh 5vw;box-sizing:border-box;" + FONT);
    panel.setAttribute("data-part", "panel");
    var card = el("div",
      "background:#1b1b1f;color:#eee;border-top:3px solid #e5484d;padding:12px 16px;" +
      "max-width:1100px;margin:0 auto;");
    card.setAttribute("data-part", "card");
    title = el("div", "color:#ff8b8b;font-weight:bold;margin-bottom:8px;");
    list = el("div", "");
    var hint = el("div", "color:#999;margin-top:10px;");
    hint.textContent = "Esc or click outside to dismiss. Fix the error and save: this clears on the next successful build.";
    card.appendChild(title);
    card.appendChild(list);
    card.appendChild(hint);
    panel.appendChild(card);
    root.appendChild(badge);
    root.appendChild(panel);
    (doc.body || doc.documentElement).appendChild(host);
    panel.addEventListener("click", function (e) {
      if (e.target === panel) dismiss();
    });
    doc.addEventListener("keydown", function (e) {
      if (e.key === "Escape") dismiss();
    });
  }

  function dismiss() {
    if (!st.failure) return;
    st.dismissed = true;
    render();
  }

  function base(p) {
    var s = String(p || "");
    var i = Math.max(s.lastIndexOf("/"), s.lastIndexOf("\\"));
    return i >= 0 ? s.slice(i + 1) : s;
  }

  function ms(n) {
    return n >= 1000 ? (n / 1000).toFixed(1) + " s" : n + " ms";
  }

  function succeeded(text) {
    st.phase = "done";
    st.text = text;
    st.stage = null;
    st.progress = null;
    st.failure = null;
    st.dismissed = false;
  }

  function apply(ev) {
    switch (ev.type) {
      case "change_detected": {
        var n = (ev.paths || []).length;
        st.phase = "change";
        st.text = "change: " + base((ev.paths || [])[0]) + (n > 1 ? " (+" + (n - 1) + ")" : "");
        break;
      }
      case "decided": {
        var d = ev.decision || {};
        if (d.tier === "overlay") { st.phase = "patching"; st.text = "overlay patch"; }
        else if (d.tier === "hot_patch") { st.phase = "patching"; st.text = "hot patch: " + (d.crates || []).join(", "); }
        else if (d.tier === "rebuild") { st.phase = "building"; st.text = "rebuild"; }
        else if (d.tier === "unchanged") {
          // The source is back to what is running: an error from a build
          // of some other source no longer describes it.
          succeeded("no change");
          st.phase = "idle";
        }
        break;
      }
      case "build_started":
        st.phase = "building";
        st.text = ev.cause === "initial" ? "building" : "rebuilding";
        st.stage = null;
        st.progress = null;
        st.pending = [];
        break;
      case "stage_started":
        st.stage = ev.stage;
        break;
      case "cargo_progress":
        st.progress = { compiled: ev.compiled, total: ev.total, current: ev.current };
        break;
      case "diagnostic":
        if (ev.diagnostic && st.pending.length < 64) st.pending.push(ev.diagnostic);
        break;
      case "build_finished": {
        if (ev.outcome === "failed") {
          st.phase = "error";
          st.text = "build failed";
          st.stage = null;
          st.progress = null;
          st.failure = { target: ev.target, error: ev.error, diagnostics: st.pending };
          st.dismissed = false;
        } else if (ev.outcome === "reloaded" || ev.outcome === "premint_refreshed") {
          succeeded("rebuilt in " + ms(ev.ms) + ", reloading");
        } else if (ev.outcome === "unchanged") {
          succeeded("rebuilt, nothing changed");
        } else {
          succeeded("ready");
        }
        st.pending = [];
        break;
      }
      case "patch_built":
        succeeded("hot patch · " + ev.redirected + " fn · " + ms(ev.ms));
        break;
      case "patch_failed":
        st.text = "no patch (" + ev.reason + "), rebuilding";
        break;
      case "overlay_pushed":
        succeeded("overlay · " + ev.sites + " site(s) · " + ms(ev.ms));
        break;
      case "sidecar_applied":
        succeeded((ev.how === "hot_patch" ? "hot patch" : "respawn") + " · " + ms(ev.ms));
        break;
      case "error":
        st.phase = "error";
        st.text = ev.source + ": " + ev.message;
        break;
    }
    render();
  }

  var ICON = { idle: "○", change: "…", patching: "↻", building: "⚙", done: "✓", error: "✗" };

  function render() {
    mount();
    var t = ICON[st.phase] + " " + st.text;
    if (st.phase === "building") {
      if (st.progress) {
        t += " " + st.progress.compiled + (st.progress.total ? "/" + st.progress.total : "");
        if (st.progress.current) t += " " + st.progress.current;
      }
      if (st.stage) t += " · " + st.stage;
    }
    badge.textContent = t;
    badge.style.color = st.phase === "error" ? "#ff8b8b" : st.phase === "done" ? "#8be28b" : "#eee";

    var show = !!st.failure && !st.dismissed;
    panel.style.display = show ? "block" : "none";
    if (!show) return;
    var f = st.failure;
    title.textContent = "Build failed" + (f.target ? " (" + f.target + ")" : "");
    while (list.firstChild) list.removeChild(list.firstChild);
    var diags = f.diagnostics || [];
    if (!diags.length) {
      var pre = el("pre", "white-space:pre-wrap;margin:0;");
      pre.textContent = f.error || "";
      list.appendChild(pre);
    }
    for (var i = 0; i < diags.length; i++) {
      var d = diags[i];
      var item = el("div", "margin:0 0 12px;");
      item.setAttribute("data-part", "diagnostic");
      var where = el("div", "color:#8fb8ff;margin-bottom:4px;");
      var loc = d.file ? d.file + (d.line ? ":" + d.line + (d.column ? ":" + d.column : "") : "") : "";
      where.textContent = loc ? loc + " — " + d.message : d.message;
      var body = el("pre", "white-space:pre-wrap;margin:0;color:#ddd;");
      body.textContent = d.rendered || d.message;
      item.appendChild(where);
      item.appendChild(body);
      list.appendChild(item);
    }
  }

  return { apply: apply, dismiss: dismiss, state: st };
}

File: crates/dev/http/tests/status_overlay.rs
//! The in-page build-state overlay renders what the dev loop sends it.
//!
//! The overlay is plain JS (it must work when the page's wasm is what
//! failed to build), so it is exercised in node against a small fake DOM
//! — enough of `document` for the narrow API the overlay uses. The
//! harness feeds it `dev-state` objects exactly as the reload stream
//! sends them and reads back what is on screen.
//!
//! Needs `node` on `PATH`. Without it the test says so and passes: the
//! end-to-end test (`crates/tools/cli/tests/wasm_hot_patch_e2e.rs`)
//! checks the same overlay in a real browser.

use std::io::Write;
use std::process::{Command, Stdio};

const HARNESS: &str = r##"
// A fake DOM: elements with children, text, style, attributes, listeners.
function El(tag) {
  this.tagName = tag; this.children = []; this.parent = null;
  this.style = {}; this.attrs = {}; this.listeners = {}; this._text = null;
}
El.prototype.appendChild = function (c) { c.parent = this; this.children.push(c); return c; };
El.prototype.removeChild = function (c) { this.children = this.children.filter(x => x !== c); return c; };
Object.defineProperty(El.prototype, "firstChild", { get() { return this.children[0] || null; } });
Object.defineProperty(El.prototype, "textContent", {
  get() { return this._text !== null ? this._text : this.children.map(c => c.textContent).join("\n"); },
  set(v) { this._text = String(v); this.children = []; }
});
El.prototype.setAttribute = function (k, v) { this.attrs[k] = v; };
El.prototype.addEventListener = function (k, f) { (this.listeners[k] = this.listeners[k] || []).push(f); };
El.prototype.attachShadow = function () { return this; };
El.prototype.fire = function (k, e) { (this.listeners[k] || []).forEach(f => f(e)); };
El.prototype.find = function (part) {
  if (this.attrs["data-part"] === part) return this;
  for (const c of this.children) { const f = c.find(part); if (f) return f; }
  return null;
};
El.prototype.findAll = function (part, out) {
  out = out || [];
  if (this.attrs["data-part"] === part) out.push(this);
  for (const c of this.children) c.findAll(part, out);
  return out;
};
const document = new El("#document");
document.body = new El("body");
document.createElement = t => new El(t);

__OVERLAY__

const o = idealystStatusOverlay(document);
const lines = require("fs").readFileSync(0, "utf8").split("\n").filter(Boolean);
const out = [];
for (const line of lines) {
  if (line.startsWith("#")) {
    const cmd = line.slice(1).trim();
    if (cmd === "esc") document.fire("keydown", { key: "Escape" });
    const host = document.body.children[0];
    const panel = host.find("panel");
    out.push({
      badge: host.find("badge").textContent,
      panel: panel.style.display !== "none",
      panelText: panel.textContent,
      diagnostics: host.findAll("diagnostic").length,
    });
    continue;
  }
  o.apply(JSON.parse(line));
}
process.stdout.write(JSON.stringify(out));
"##;

fn node() -> Option<String> {
    Command::new("node").arg("--version").output().ok()?.status.success().then(|| "node".into())
}

fn run(script: &[&str]) -> Option<Vec<serde_json::Value>> {
    let node = node()?;
    let harness = HARNESS.replace("__OVERLAY__", dev_http::STATUS_OVERLAY_JS);
    let mut child = Command::new(node)
        .args(["-e", &harness])
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .expect("spawn node");
    child.stdin.take().unwrap().write_all(script.join("\n").as_bytes()).unwrap();
    let out = child.wait_with_output().unwrap();
    assert!(out.status.success(), "{}", String::from_utf8_lossy(&out.stderr));
    Some(serde_json::from_slice(&out.stdout).expect("the harness prints JSON"))
}

fn skip() {
    eprintln!("status_overlay: `node` is not on PATH; skipped (the CLI E2E covers the overlay)");
}

const BUILD: &str = r#"{"v":1,"seq":1,"at_ms":0,"type":"build_started","target":"web","cause":"save","folded":0}"#;
const STAGE: &str = r#"{"v":1,"seq":2,"at_ms":1,"type":"stage_started","target":"web","stage":"cargo"}"#;
const PROGRESS: &str = r#"{"v":1,"seq":3,"at_ms":2,"type":"cargo_progress","target":"web","compiled":212,"total":480,"current":"idea-ui"}"#;
const DIAG: &str = r#"{"v":1,"seq":4,"at_ms":3,"type":"diagnostic","target":"web","diagnostic":{"level":"error","message":"mismatched types","code":"E0308","file":"src/app.rs","line":12,"column":9,"rendered":"error[E0308]: mismatched types\n  --> src/app.rs:12:9\n"}}"#;
const FAILED: &str = r#"{"v":1,"seq":5,"at_ms":4,"type":"build_finished","target":"web","outcome":"failed","error":"cargo exited with exit status: 101","ms":3100}"#;
const PATCHED: &str = r#"{"v":1,"seq":6,"at_ms":5,"type":"patch_built","target":"web","files":["src/app.rs"],"crates":[],"redirected":3,"steps":[],"skipped":[],"bytes":10,"ms":446}"#;

#[test]
fn the_overlay_shows_progress_then_the_error_and_clears_on_success() {
    let Some(frames) = run(&[BUILD, STAGE, PROGRESS, "#", DIAG, FAILED, "#", PATCHED, "#"]) else {
        return skip();
    };
    // Building: the badge carries the progress and the stage.
    let building = &frames[0];
    assert_eq!(building["badge"], "⚙ rebuilding 212/480 idea-ui · cargo", "{building}");
    assert_eq!(building["panel"], false);

    // Failed: the panel is up with the diagnostic's file:line and its
    // rendered message.
    let failed = &frames[1];
    assert_eq!(failed["panel"], true, "{failed}");
    assert_eq!(failed["diagnostics"], 1);
    let text = failed["panelText"].as_str().unwrap();
    assert!(text.contains("Build failed (web)"), "{text}");
    assert!(text.contains("src/app.rs:12:9 — mismatched types"), "{text}");
    assert!(text.contains("error[E0308]: mismatched types"), "{text}");
    assert_eq!(failed["badge"], "✗ build failed");

    // The next success clears it.
    let fixed = &frames[2];
    assert_eq!(fixed["panel"], false, "{fixed}");
    assert_eq!(fixed["badge"], "✓ hot patch · 3 fn · 446 ms");
}

#[test]
fn escape_dismisses_the_error_until_the_next_failure() {
    let Some(frames) = run(&[BUILD, DIAG, FAILED, "#", "#esc", BUILD, DIAG, FAILED, "#"]) else {
        return skip();
    };
    assert_eq!(frames[0]["panel"], true);
    assert_eq!(frames[1]["panel"], false, "Esc dismisses");
    assert_eq!(frames[2]["panel"], true, "a new failure shows again");
}

#[test]
fn a_failure_with_no_diagnostic_shows_the_error_itself() {
    let Some(frames) = run(&[BUILD, FAILED, "#"]) else { return skip() };
    assert_eq!(frames[0]["diagnostics"], 0);
    assert!(
        frames[0]["panelText"].as_str().unwrap().contains("cargo exited with exit status: 101"),
        "{}",
        frames[0]
    );
}

#[test]
fn the_overlay_is_part_of_the_served_reload_script() {
    let script = dev_http::reload_script_tag("/__idealyst/reload");
    assert!(script.contains("function idealystStatusOverlay(doc)"));
    assert!(script.contains(r#"es.addEventListener("dev-state""#));
    assert!(!script.contains("__STATUS_OVERLAY__"));
}

/// Undoing the edit that broke the build returns the source to what is
/// running; the error no longer describes it, so it goes.
#[test]
fn saving_back_to_the_running_source_clears_the_error() {
    const UNCHANGED: &str = r#"{"v":1,"seq":9,"at_ms":9,"type":"decided","target":"web","decision":{"tier":"unchanged"}}"#;
    let Some(frames) = run(&[BUILD, DIAG, FAILED, "#", UNCHANGED, "#"]) else { return skip() };
    assert_eq!(frames[0]["panel"], true);
    assert_eq!(frames[1]["panel"], false, "{}", frames[1]);
    assert_eq!(frames[1]["badge"], "○ no change");
}
